src/main/java/com/czertainly/core/service/cmp/profiles/Mobile3gppProfileContext.java

```diff
@@ -0,0 +1,208 @@
+     *      is optional.</p>
+     *
+     */
+    @Override
+    public void validateResponse(PKIMessage response) throws CmpProcessingException {
+        ASN1OctetString tid = response.getHeader().getTransactionID();
+        switch (response.getBody().getType()) {
+            case PKIBody.TYPE_INIT_REP:
+            case PKIBody.TYPE_CERT_REP:
+            case PKIBody.TYPE_KEY_UPDATE_REP:
+                if(response.getExtraCerts() == null || response.getExtraCerts().length==0) {
+                    throw new CmpProcessingException(tid,
+                            PKIFailureInfo.badDataFormat, "field 'extraCerts' is null or empty");
+                }
+                break;// do something
+            default:
+                throw new CmpProcessingException(tid,
+                        PKIFailureInfo.badDataFormat, "only CRMF-based message can be validated");
+        }
+    }
+
+    @Override
+    public ProtectionMethod getProtectionMethod() throws CmpConfigurationException {
+        return getProfile().getRequestProtectionMethod();
+        //return ProtectionMethod.SHARED_SECRET;
+    }
+
+    @Override
+    public ProtectionStrategy getProtectionStrategy() throws CmpBaseException {
+        ProtectionMethod czrtProtectionMethod = getProfile().getResponseProtectionMethod();
+        //return ProtectionMethod.SHARED_SECRET;
+        switch (czrtProtectionMethod){
+            case SIGNATURE:
+                return new SingatureBaseProtectionStrategy(this,
+                    requestMessage.getHeader().getProtectionAlg(), certificateKeyService);
+            case SHARED_SECRET:
+                byte[] salt = CertUtil.generateRandomBytes(20);//precist z db
+                int iterationCount = 1000;//precist z db
+                return new PasswordBasedMacProtectionStrategy(this,
+                        requestMessage.getHeader().getProtectionAlg(),
+                        getSharedSecret(), salt, iterationCount);
+            default:
+                throw new CmpConfigurationException(requestMessage.getHeader().getTransactionID(),
+                        PKIFailureInfo.systemFailure,
+                        "wrong config3gppProfile: unknown type of protection strategy, type="+czrtProtectionMethod);
+        }
+    }// pri vyberu
+
+    @Override
+    public byte[] getSharedSecret() {
+        /* senderKID field MUST hold an identifier
+         *    that indicates to the receiver the appropriate shared secret
+         *    information to use to verify the message */
+        ASN1OctetString senderKID = requestMessage.getHeader().getSenderKID();//muze byt pouzit pro dohledani v db
+        return profile.getSharedSecret().getBytes();
+    }
+
+    @Override
+    public List<RequestAttributeDto> getClientOperationAttributes(boolean isRevoke) {
+        return (isRevoke) ? revokeAttributes : issueAttributes;
+    }
+
+    @Override
+    public boolean dumpSinging() {

```

Comment:
I do not understand this `dumpSinging` method...

Reply:
It will help troubleshooting (content of extraCerts field) during 3gpp testing (the code will be removed after testing).